src/dso/dso_transport.cpp

```diff
@@ -0,0 +1,552 @@
+    std::vector<uint8_t> buf(length + sizeof(uint16_t));
+    uint16_t             lengthInBigEndian = htons(length);
+    int                  writeLen;
+
+    VerifyOrExit(mState == State::kConnected);
+
+    otbrLogDebug("Sending a message with length %" PRIu16, length);
+
+    memcpy(buf.data(), &lengthInBigEndian, sizeof(lengthInBigEndian));
+    VerifyOrExit(length == otMessageRead(aMessage, 0, buf.data() + sizeof(lengthInBigEndian), length),
+                 otbrLogWarning("Failed to read message data"));
+    writeLen = mbedtls_net_send(&mCtx, buf.data(), buf.size());
+    if (writeLen < length)
+    {
+        otbrLogWarning("Failed to send DSO message: %d", writeLen);
+    }
+    HandleMbedTlsError(writeLen);
+
+    // TODO: May need to keep sending until all the data is sent
+exit:
+    return;
+}
+
+void DsoAgent::DsoConnection::HandleReceive(void)
+{
+    int     readLen;
+    uint8_t buf[kRxBufferSize];
+
+    VerifyOrExit(mState == State::kConnected);
+
+    while (true)
+    {
+        if (mNeedBytes)
+        {
+            readLen = mbedtls_net_recv(&mCtx, buf, std::min(sizeof(buf), mNeedBytes));
+
+            if (readLen <= 0)
+            {
+                HandleMbedTlsError(readLen);
+                VerifyOrExit(readLen != 0);
+                VerifyOrExit(readLen != MBEDTLS_ERR_SSL_WANT_READ && readLen != MBEDTLS_ERR_SSL_WANT_WRITE);
+                otbrLogWarning("Failed to receive message: %d", readLen);
+                ExitNow();
+            }
+            SuccessOrExit(otMessageAppend(mPendingMessage, buf, readLen));
+            mNeedBytes -= readLen;
+
+            if (!mNeedBytes)
+            {
+                otPlatDsoHandleReceive(mConnection, mPendingMessage);
+                mPendingMessage = nullptr;
+            }
+        }
+        else
+        {
+            assert(mLengthBuffer.size() < sizeof(uint16_t));
+            assert(mPendingMessage == nullptr);
+
+            readLen = mbedtls_net_recv(&mCtx, buf, sizeof(uint16_t) - mLengthBuffer.size());
+
+            if (readLen <= 0)
+            {
+                HandleMbedTlsError(readLen);
+                VerifyOrExit(readLen != 0);
+                VerifyOrExit(readLen != MBEDTLS_ERR_SSL_WANT_READ && readLen != MBEDTLS_ERR_SSL_WANT_WRITE);
+                otbrLogWarning("Failed to receive message: %d", readLen);
+                ExitNow();
+            }
+            for (int i = 0; i < readLen; ++i)
+            {
+                mLengthBuffer.push_back(buf[i]);
+            }
+            if (mLengthBuffer.size() == sizeof(uint16_t))
+            {
+                mNeedBytes      = mLengthBuffer[0] << 8 | mLengthBuffer[1];
+                mPendingMessage = otIp6NewMessage(otPlatDsoGetInstance(mConnection), nullptr);
+                mLengthBuffer.clear();
+                VerifyOrExit(mPendingMessage != nullptr);
+            }
+        }
+    }
+
+exit:
+    return;
+}
+
+void DsoAgent::DsoConnection::HandleMbedTlsError(int aError)
```

Comment:
Log the error

Reply:
The error has been logged at call sites.